src/sdg/organisaties/views/catalogi.py

```diff
@@ -23,56 +21,42 @@ def get(self, request, *args, **kwargs):
-            for cat in role.get_catalogs(reference=False).select_related(
-                "referentie_catalogus"
-            ):
-                ref_cat = cat.referentie_catalogus
-                if (
-                    ref_cat.user_is_redacteur(self.request.user)
-                    and ref_cat.lokale_overheid == self.object
-                ):
-                    setattr(ref_cat, "area_and_products", deepcopy(area_and_products))
-                setattr(cat, "area_and_products", deepcopy(area_and_products))
-                catalogs.append(cat)
-            for theme in themes.filter(upn__generieke_producten__isnull=False):
-                area = theme.informatiegebied.informatiegebied
-                for catalog in catalogs:
-                    reference_catalog = catalog.referentie_catalogus
-                    products = Product.objects.filter(
-                        referentie_product__generiek_product__upn__in=theme.upn.all(),
-                        catalogus=catalog,
-                    )
-                    reference_products = Product.objects.filter(
-                        generiek_product__upn__in=theme.upn.all(),
-                        catalogus=reference_catalog,
-                    )
-                    catalog.area_and_products[area].update(
-                        products
-                        | reference_products.exclude(specifieke_producten__in=products)
-                    )
-                    if getattr(reference_catalog, "area_and_products", None):
-                        reference_catalog.area_and_products[area].update(
-                            reference_products
-                        )
+        catalogs = ProductenCatalogus.objects.specific_catalogs(
+            municipality=self.object
+        ).annotate_area_and_products()
+        for catalog in catalogs:
+            reference_catalog = catalog.referentie_catalogus
 
-        context["catalogs"] = catalogs
+            products = (
+                Product.objects.filter(catalogus=catalog)
+                .most_recent()
+                .annotate_name_and_area()
+                .select_generic()
+                .exclude(area__isnull=True)
+            )
+            reference_products = (
+                Product.objects.filter(catalogus=reference_catalog)
+                .most_recent()
+                .annotate_name_and_area()
+                .select_generic()
+                .exclude(area__isnull=True)
+            )
+
+            intersected_products = list(
+                products | reference_products.exclude(specifieke_producten__in=products)
+            )
 
+            for product in intersected_products:
+                catalog.area_and_products[product.area].add(product)
+
+            reference_areas = getattr(reference_catalog, "area_and_products", None)
+            if reference_areas:
+                for product in reference_products:
+                    reference_areas[product.area].add(product)
```

Comment:
```suggestion
            reference_catalog.reference_areas = defaultdict(set)
            for product in reference_products:
                reference_catalog.reference_areas[product.area].add(product)
```

The typical case only has 1 catalog in this view, but if you want, you can still "cache" the reference catalogs.

With these code changes, the same amount of queries are done in the typical case, without any manager function that is only used in this view.